docs/Dialects/Verif.md

```diff
@@ -4,6 +4,322 @@ This dialect provides a collection of operations to express various verification
 
 [TOC]
 
+## Contracts
+
+Formal contracts are a key building block of the Verif dialect to help make formal verification scale to larger designs and deep module hierarchies.
+Contracts describe what a circuit expects from its inputs (`verif.require`) and what it guarantees its output to be (`verif.ensure`).
+The `verif.contract` op can be inserted into an SSA edge similar to an `hw.wire`, where the contract simply passes its operands through to its results.
+During formal verification the contract results are replaced by symbolic values that uphold the guarantees described in the contract's body.
+
+These contracts can then be used in two key ways:
+
+1.  A contract can be _verified_ to hold by turning `require`s into `assume`s and `ensure`s into `assert`s.
+    Doing so verifies that a circuit upholds the contract by placing asserts on the values it produces, and by placing assumes on the input values the circuit sees.
+    In a nutshell, this checks that, assuming the inputs honor the contract, the circuit also upholds the contract.
+    This check can be done very efficiently by creating `verif.formal` ops to verify each contract.
+
+2.  A contract can be _assumed_ to hold by turning `require`s into `assert`s and `ensure`s into `assume`s.
+    Doing so verifies that the inputs fed into a circuit uphold the contract, such that the outputs can be assumed to have the promised values.
```

Comment:
I'm not entirely sure I'm a fan of this language. We're not assuming a contract, rather we're using it as an abstraction for the module during the verification of an instance.

Reply:
Yeah good point. I was trying to make the contract feel like an equivalent to asserts and assumes, where you just assert and assume a higher-level concept instead of individual signals. Especially since we don't have any modules and instances tied to the contracts. It feels more like a local abstraction that you can formulate which then gets substituted in.